### ISSUE TITLE
iptables commands should wait on xtables lock

### ISSUE THREAD

**Issue by serbrech** [Azure/azure-container-networking#252]
**Is this a request for help?**: 
No
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one): 
ISSUE
---

**Which release version?**: 
master

---

**Which component (CNI/IPAM/CNM/CNS)**: 
CNS, Network Policy Manager, and really everything touching iptables. 

---

**Which Operating System (Linux/Windows)**: 
Linux

---

**Which Orchestrator and version (e.g. Kubernetes, Docker)**

---

**What happened**: 
Other containers and processes are updating iptables : `kube-router`, `kube-proxy` for example. They rely on other processes honoring the xtables lock from iptables.  
From the iptables documentation : 

```
-w, --wait
Wait for the xtables lock. To prevent multiple instances of the program from running concurrently, an attempt will be made to obtain an exclusive lock at launch. 
By default, the program will exit if the lock cannot be obtained. This option will make the program wait until the exclusive lock can be obtained.
```

See examples of iptables usage that handle this correctly: 
- [go-iptables](https://github.com/coreos/go-iptables/blob/47f22b0dd3355c0ba570ba12b0b8a36bf214c04b/iptables/iptables.go#L363) used by kube-router
- [iptables.go](https://github.com/kubernetes/kubernetes/blob/a27adf1612d267b4460b420cec544ba4e3786cad/pkg/util/iptables/iptables.go#L431) in kubernetes.

note that `iptables-restore` [needs special treatment](https://github.com/kubernetes/kubernetes/blob/a27adf1612d267b4460b420cec544ba4e3786cad/pkg/util/iptables/iptables.go#L372) as it does not support `--wait` flag

---

**What you expected to happen**:
CNI should play nice with other network plugins and processes that use iptables
---

**How to reproduce it** (as minimally and precisely as possible):

I will try to come up with a minimal reproduction.
Idea : run an `iptables` command with `--wait` in a loop on a node where cni is configured and see it fail with `Resource temporarily unavailable.` error because the file was opened without a lock being taken.

---

**Anything else we need to know**:

linked issues : 
- https://github.com/Azure/AKS/issues/607
- https://github.com/cloudnativelabs/kube-router/issues/506

people : @marrobi 

---

**Comment by moomzni** [comment]
We're experiencing this issue running on AKS v1.10.5 clusters with Kube-router.

We are seeing kube-router crash and restart every 5 minutes with error messages similar to the below:

```E1106 17:50:41.949477 1 network_policy_controller.go:167] Error syncing network policy for the update to pod: default/emis-check-1541526060-rrqtq Error: Aborting sync. Failed to sync network policy chains: Failed to run iptables command: running [/sbin/iptables -t filter -A KUBE-NWPLCY-4FCZBZLNJQ5CNI23 -m comment --comment rule to ACCEPT traffic from specified CIDR's to dest pods selected by policy name: allow-randomrule namespace random -m set --set KUBE-SRC-AGSAGXMZ25A3C3XM src -p TCP --dport 443 -d 45.35.33.0/24 -j ACCEPT --wait]: exit status 4: --set option deprecated, please use --match-set
iptables: Resource temporarily unavailable.```

You'll note that kube-router is using the `--wait` flag and we are mounting the `/run/xtables.lock` into the container so it should play nicely with neighbours...

Doing a brief analysis it looks like the AzureCNI does mount the /var/run directory so should be honouring `--wait` commands if used; but both kube-proxy and kube-svc-redirector are not (assuming both have some relationship to iptables).

**Comment by marrobi** [comment]
I'm not sure how Azure CNI uses iptables, but dosee references to iptables commands in a number of places, such as https://github.com/Azure/azure-container-networking/blob/45914be23d765486e29ac0190a6a0a0c9383e73a/platform/os_linux.go#L79

There does not seem to be a wait argument on the command?

**Comment by tamilmani1989** [comment]
Azure CNI never calls this function SetOutboundSNAT. Its used by CNS not CNI.
https://github.com/Azure/azure-container-networking/blob/45914be23d765486e29ac0190a6a0a0c9383e73a/platform/os_linux.go#L78 

Azure CNI is not calling any iptable commands.

**Comment by saiyan86** [comment]
Fixed in #315
